feat(edit-column): initial implementation based on prototype

web-padawan opened the pull request:
Fixes #3 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-grid-pro/8)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/vaadin-grid-pro?pullRequest=8) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/vaadin-grid-pro?pullRequest=8) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/vaadin-grid-pro?pullRequest=8) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/vaadin-grid-pro?pullRequest=8) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/vaadin-grid-pro?pullRequest=8) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/vaadin-grid-pro?pullRequest=8) it.</sub>

web-padawan commented:
Dismissed the second comment from the CLA assistant, as the CLA actually signed.

tomivirkki commented on src/vaadin-pro-grid-inline-editing-mixin.html:
```diff
@@ -0,0 +1,206 @@
+      this._cancelStopEdit();
+
+      const columnTreeLevel = this._columnTree.length - 1;
+      const cols = this._columnTree[columnTreeLevel].filter(this._isEditColumn);
```
I think you'll need to sort these by column `_order` which is the current effective order (drag-reordering doesn't actually reorder the column elements in the DOM). Also, some columns might be hidden so those should be filtered out. [Example](https://github.com/vaadin/vaadin-grid/blob/805e1158795f3563238d49c058dbf91c29b3e4ca/src/vaadin-grid-column-resizing-mixin.html#L44)

tomivirkki commented on src/vaadin-pro-grid-inline-editing-mixin.html:
```diff
@@ -0,0 +1,206 @@
+        this.notifyPath('items.' + index + '.' + path, value);
+      });
+
+      this.addEventListener('click', e => {
```
Yes, doubleclick should be the default

web-padawan replied:
Done.

tomivirkki commented on src/vaadin-pro-grid-edit-column.html:
```diff
@@ -0,0 +1,190 @@
+        this.__editStateTemplate = this._prepareTemplatizer(this.$.editStateTemplate);
+        // needed to override the dataHost correctly in case internal template is used.
+        this.__editStateTemplate.templatizer.dataHost = this;
+        this.__editStateRenderer = this.constructor._getEditorRenderer();
```
As you scroll the grid while the cell is in edit mode, the same editor component will appear on an arbitrary row with the content of the cell being actually edited

web-padawan replied:
Discussed in person and it is OK for now to exit edit state when scrolled out of the viewport.

web-padawan replied:
Done.

web-padawan commented on src/vaadin-pro-grid-edit-column.html:
```diff
@@ -0,0 +1,190 @@
+See <a href="https://vaadin.com/license/cval-3">the website</a> for the complete license.
+-->
+
+<link rel="import" href="../../polymer/lib/elements/dom-if.html">
```
Removed.

tomivirkki commented on src/vaadin-pro-grid-edit-column.html:
```diff
@@ -0,0 +1,164 @@
+See <a href="https://vaadin.com/license/cval-3">the website</a> for the complete license.
+-->
+
+<link rel="import" href="../../vaadin-grid/vaadin-grid-column.html">
```
Should import from src

web-padawan replied:
Done.

web-padawan commented on src/vaadin-pro-grid-inline-editing-mixin.html:
```diff
@@ -0,0 +1,206 @@
+            value: value
+          },
+          bubbles: true,
+          composed: true
```
Done.

web-padawan commented on demo/pro-grid-basic-demos.html:
```diff
@@ -6,13 +6,54 @@
+          </vaadin-pro-grid-edit-column>
+          <vaadin-grid-column header="Last name (view)">
+            <template>
+              <span style="color: #999;">[[item.name.last]]</span>
```
Will make more sense once we have `editorRenderer`, so let me improve it in #14

tomivirkki commented on src/vaadin-pro-grid-inline-editing-mixin.html:
```diff
@@ -0,0 +1,239 @@
+        nextIdx = e.shiftKey ? index - 1 : index + 1;
+      }
+
+      // tab: move right / left
```
Seems that the edit mode is lost on tab press. Worked with a previous revision.

web-padawan replied:
Yep, that is by design:

> Hitting tab while editing a cell moves the focus in unedited mode to the cell to the right

It works the same way in Google Spreadsheet, I mentioned this in the morning.

web-padawan commented:
Based on the discussion with @tomivirkki, @platosha and @Peppe, the keyboard navigation behavior has been slightly changed to introduce less breaking changes:

1. Added `preserveEditMode` flag (default: false) to activate edit mode on the next focused cell

2. Added `allowEnterRowChange` flag (default: false) to move focus to the next / previous row on <kbd>Enter</kbd> / <kbd>Shift</kbd> + <kbd>Enter</kbd>

Note that making <kbd>Tab</kbd> behavior in edit mode opt-in also is trickier, due to how it is currently implemented `KeyboardNavigationMixin`. Submitted #17 to handle that in a separate PR

web-padawan commented:
@limonte I'm unable to reproduce Esc issue with my virtual machine (EdgeHTML 17.17134)
Looks like it only happens after upgrading to 18. Could you please submit a separate issue for that?

limonte commented:
> Looks like it only happens after upgrading to 18. Could you please submit a separate issue for that?

Done, #18
